### PR TITLE
[FIX] account: reconciliation is creating wrong payments

### PR DESCRIPTION
**Pull request by wtaferner**
A partner plus a simple positive or negative amount is not enough to decide whether we
have to create a supplier or a customer payment, so we do rely on the
accounts used in the move lines to be created.

**Description of the issue/feature this PR addresses:**
Be more precise to keep a consistent logic for ```account.payment``` otherwise we have just wrong payments registered (a negative payment could be a refund of a receivable)

**Current behavior before PR:**
Wrong payments are created through bank statement reconciliation

**Desired behavior after PR is merged:**
Choose the right ```partner_type``` based on accounts and block certain use cases and give some feedback

@qdp-odoo @nim-odoo
May I ask you to review and merge this for me? I wonder nobody complained yet as this is happening for more than three years now and should be a very common use case.

OPW-1910807

Info: @wt-io-it

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/account/models/account_bank_statement.py [hunk 1]
```diff
@@ -915,11 +925,29 @@ def process_reconciliation(self, counterpart_aml_dicts=None, payment_aml_rec=Non
...
-                if partner_id:
-                    if total < 0:
-                        partner_type = 'supplier'
-                    else:
-                        partner_type = 'customer'
+                # Check based on account type whether we have a payment for receivable (customer) or payable (supplier) accounts and fallback in case none of them are used
+                if partner_id and partner_id != self.env.user.company_id.partner_id:
+                    for account in partner_accounts:
+                        block = False
+                        if account.user_type_id.id == self.env.ref('account.data_account_type_payable').id:
```
Comment (fmdl) [inline addons/account/models/account_bank_statement.py]: you could put `account_type_payable = self.env.ref('account.data_account_type_payable')` before the loop for, to improve the speed.

File: addons/account/models/account_bank_statement.py [hunk 2]
```diff
@@ -915,11 +925,29 @@ def process_reconciliation(self, counterpart_aml_dicts=None, payment_aml_rec=Non
...
+                            if partner_type == 'customer':
+                                block = True
+                            partner_type = 'supplier'
+                        elif account.user_type_id.id == self.env.ref('account.data_account_type_receivable').id:
```
Comment (fmdl) [inline addons/account/models/account_bank_statement.py]: you don't need to add .id, `account.user_type_id == self.env.ref('account.data_account_type_receivable')` is enought

File: addons/account/models/account_bank_statement.py (hunk 1 above)
Comment (wtaferner) [reply addons/account/models/account_bank_statement.py]: @fmdl 
Jep, I thought about it, but was concentrated on the logical and accounting part...will do now...thx

**Comment by wtaferner** [comment]
@nim-odoo @qdp-odoo 
Failed for a phantom JS test which seems not to find the companies transfer account in the backend. I choose to change the receivable account as the PoS is more like a on the fly action and no partner is given. We could eventually use a "Public User" as a partner, but this is getting to much for stable already. 

Should we follow up this approach and fix the tests or do you think it is better to add a context to tell the newly created blocker that it is exceptionally allowed to use a receivable account without a partner here?

Have I missed anything else? Functionally the solution is working fine from my PoV.

**Comment by nim-odoo** [comment]
Can you please detail the use case to reproduce the issue? Thanks

**Comment by wtaferner** [comment]
@nim-odoo 
Sorry, I thought it was clear...

It is about reconciliation of bank statement lines...in case you have a negative amount on the statement line and you like to create a payment from the line it will always create a supplier payment in case you book it on the statement with a receivable account as it is a refund of a customer instead of a payment to a supplier.

So the interface does assume a complete wrong business case from an correctly encoded accounting move. Got it?

**Comment by wtaferner** [comment]
@nim-odoo 
Are you clear about the use case? I just need some feedback how we resolve the false positive in the test. The topic and wrong behavior itself is clear IMO.

**Comment by wtaferner** [comment]
@qdp-odoo @nim-odoo 
Fixed the tests for PoS and would be happy if you could take a closer look as the issue is real and real bad ;-) 
The solution I propose is maybe a change of behavior in the PoS, but actually smarter than it was implemented before IMO as in reality there is no payable and receivable account for an unknown partner, right?

Basically the transfer account should be defined by any installation, so it will only change the intermediate account for direct payment without a partner linked and the best is that we do have proper payments created from the statement reconciliation after merging this one :tada:

**Comment by nim-odoo** [comment]
@wtaferner it's still on my TODO list, but clearly not something I can review in a few minutes.

**Comment by wtaferner** [comment]
@nim-odoo 
Thx for the feedback, as long as I know that it is not forgotten or closed (i.e. for https://github.com/odoo/odoo/pull/28129, the OPW was automatically closed today although I think a closer look is needed - no response was given or call was scheduled -  and I still challenge both of you...please be so kind and reopen), I am more or less patient (depending on when it is investigated and properly discussed, especially with things having an impact on standard accounting)

**Comment by wtaferner** [comment]
@nim-odoo 
Unclear? 

> It is about reconciliation of bank statement lines...in case you have a negative amount on the statement line and you like to create a payment from the line it will always create a supplier payment in case you book it on the statement with a receivable account as it is a refund of a customer instead of a payment to a supplier.

**Comment by wtaferner** [comment]
@gdgellatly 
Maybe you could take a look and give your opinion. I hope you understand my issue and the fix here (for me it is glass clear, but it must not be true for the rest of the world and you are on the other end of it, so would be a good validation :wink: )

**Comment by nim-odoo** [comment]
So here is a step-by-step reproduction scenario:

- Create a bank statement for a negative amount, e.g. -100.
- Click 'Reconcile'.
- Create a counterpart line for this bank statement, but specify receivable account.

The payment created has a `payment_type` set as `supplier`, although a receivable account is used for customers.

This use case can arise in case of a credit note to a customer: the reimbursement should create a customer payment, not a supplier payment.

I'll check this with @qdp-odoo because to me the diff as it is now is way too restrictive for what you are trying to solve.

**Comment by wtaferner** [comment]
@nim-odoo 
Thank you, about the restrictions I try to summarize what I try to avoid

- **Mixing receivable and payable in one move.** 
This should never happen IMO, because we can not determine for `account.payment` which use case is given and I am not sure whether this would be allowed in accounting.
- **Posting a receivable or payable account without a partner** 
maybe too restrictive, but I have the opinion that the accountant should put the amount on another intermediary account which can be reconciled instead of having an undefined receivable/payable move line which is not possible in other accounting software afaik and keep it there for reference until the partner is determined and posted accordingly)

Not sure which use cases would need to loose these restrictions and would be valid in accounting. But if we can not name them I would suggest to fix it like that and let the users find the use case otherwise a lot of bad stuff is happening I guess and someone needs to fix it in every company :wink:

**Comment by nim-odoo** [comment]
@wtaferner we would rather go for a simplified version of of this patch, see https://github.com/odoo/odoo/pull/30232

Still unsure about the `payment_type`, though.

**Comment by wtaferner** [comment]
@nim-odoo 
Unclear what and why you challenge my constraints, of course we can optimize the implementation, but the patch has a target and it is fixing bad reconciliation behavior. So, please elaborate why you think my constraints are not valid otherwise we do just half of the job and this sucks :wink: 

I have evidence that the payment type is assumed wrongly for several situations and other situations should just not happen, so the minimum would be to address each of them here publically instead of saying that you would rather go for a simplified version.

What do you prefer, the risk that someone is complaining of a corner case or hundreds of databases messed up with stupid account moves b/c of reconciliation allowing that (reconciliation = user tool to do the job more easily and quickly which is prone to errors and fixing them is not really fun)?

**Comment by nim-odoo** [comment]
> @nim-odoo
> So, please elaborate why you think my constraints are not valid otherwise we do just half of the job and this sucks 

It is usually out of the scope of bug fixing to add constrains, whatever they are.

**Comment by wtaferner** [comment]
@nim-odoo @gdgellatly 
It is all about the user...this fix was done on real errors of users and they were not practicants but more or less experienced accountants just relying too much on the seducing interface or just did not pay enough attention and my constraints are only addressing accounting constraints.

So I am perfectly fine if an international accountant would say, YES, I do normally post these things manually exactly like that in one account move and it works perfectly in my software without any side-effects. It is all about the payable and the receivable account type.

...and of course the wrong payment type which is basically an Odoo topic...

**Comment by gdgellatly** [comment]
@wtaferner I am an accountant, and I would never post a move like that. Not automatically, or manually for that matter. In the cases they arise by surprise the ordinary process is something like.

Debtors Balance $1,000,000
Creditors Balance $800,000
Payment $200,000

1. Leave balance open on bank statement and refer to accountant.
2. Work out how much of creditor invoices is paying debtors invoices (manually). Simple in this case, but rarely is*. $800,000
3. Journal Entry - DR Creditors, CR Debtors (Lets call it MISC1)
4. Reconcile MISC1 against Creditors Invoices.
5. Reconcile Open Balance and Debtors Invoices against Debtors Invoices.

So for me to do it in 1 interface cleanly, firstly you would want some kind of warning of what was being done. Secondly, in that method you would want to create 2 separate reconciliations to do it correctly and also need to be able to leave 2 open balances (see *). Otherwise, as I say, I prefer the constraints.

*The reason it is rarely that simple, and also the reason contra payments are usually pretty rare is because it requires usually at least 4 people agreeing on amounts, disputes, timings, payment dates etc, plus both systems supporting it and gives too much power to the person paying to simply pay and include disputed invoices in the offset. But also, if you are agreeing a contra payment amount, then you are effectively reconciling twice, so you would actually prefer to do the offset before the payment arrived or was sent (step 2).

**Comment by wtaferner** [comment]
@gdgellatly 
Thank you for confirming that you would never do post such moves and no, the argument (not said by you) that accountants do not need help b/c they are very serious about doing their business does not help as they are users and time is often crucial, so if you have a lot of open positions you want to book instead of keeping open until the invoice is created (prepayment of customers and sometimes paybacks which is of course a no go without a refund invoice, but happens in reality with all the payment gateways and methods from time to time).

About your surprise example...contra payment reconciliation is a complete different topic and I do agree that this is nothing to be solved by now in the UI or could be handled for example as you have written and I would not want to make it possible for your stated reasons and it is anyway already a constraint. 

There are more constraints I ask myself why they are there (like mixing an anticipated payment/bank move  and an open invoice is impossible directly and time consuming - maybe already possible in 12.0...not tested/experienced yet).

I  just want to constraint stupid moves as listed in the bank statement and give some useful default logic that does not lead to wrong posted moves and payments (wrong type) easily.

@qdp-odoo 
Any opinion or thoughts you want to share publicly?
I guess @nim-odoo did already consult you, but as said, I really think the constraints are playing a big role to fix this issue. 
Of course, you guys decide what you merge in the end, but I feel it is a contribution saving time and money for all of us later. Well, depending if there are hardcore accountants who you are afraid of if you do constraints like I want to have and will be able to force you to have more work yourself based on this :wink:

**Comment by wtaferner** [comment]
> It is usually out of the scope of bug fixing to add constrains, whatever they are.

Hmm, even if it is something in accounting which is basically a no go? Payables and Receivables are sensitive afaik and must always have a dedicated partner and need to be external of course.

**Comment by qdp-odoo** [comment]
Hello @wtaferner , I'm just passing by to give my opinion on your wannabe constraints, I didn't check the code but we discussed with @nim-odoo and are aligned together on what should be the patch (so far :-))
>Mixing receivable and payable in one move.

say you encoded wrongly a negative amount for a bank statement line in the payable account, but it was acutally a customer refund. You want to make a misc operation from payable to receivable. This constraint cannot be on the move. But I think we already have (had?) a constraint within the bank reconciliation widget forbidding to mix receivable/payable items for the same bank statement line... So it's at a higher level but the idea is the same. 

>Posting a receivable or payable account without a partner

not a good idea for any POS accounting entries, where you don't record the customer everytime. We are also working on some "suspense account" feature where, whennever you encode bank transactions for which you cannot clearly identify the partner yet, you'll let the money on the receivable/payable account without any partner and with a checkbox to remind you to fullfill the transaction later. This would help at least to have a correct bank account balance in the meantime.

>Posting a move to a receivable or payable account where the own company is the partner

For this one, I don't have any idea. But I cannot see any harm it would do to do so. So I'm no in favor of adding a constraint that could potentially block someone's use case for something that seems, not an operation you do every day?

hope that helps :-)

**Comment by wtaferner** [comment]
@qdp-odoo 
Hmm, thank you for posting your opinion / explanations. 

So, basically I give up now as I kind of feel you do not want to see an issue where I think there is a big hole in terms of keeping a clean accounting with payable/receivable and payment types (let's see if the implementation of @nim-odoo will do a good job...yay, great thing b/c now I even need to rework the constraints which do not fit after merge to resolve the conflict).
 
The point of sale indeed would need the change to the "Transfer Account" (in the patch) which would be feasible, but the work on the "Suspense Account" is more or less the same (besides a contextual issue I guess) as you park the payment on a intermediary account without a partner and can reconcile later.
Currently this is all not happening or the user is dancing with Murphy and you need to deal with it. 

So, dear Odoo community (or is it just Enterprise?) have fun until 13.0 (hopefully) and feel free to enjoy my patch and save time and money for yourself and your customers.

Donations are accepted :stuck_out_tongue_winking_eye:

**Comment by qdp-odoo** [comment]
ok, so I guess the fix landed at https://github.com/odoo/odoo/commit/c5eeae0d5a4a697ea038591b3aa64f9d6e3c20af now closes this issue

Thanks for the contribution

**Pull request closed without merging**